MDAnalysisData/base.py

```diff
@@ -94,6 +96,59 @@ def __setstate__(self, state):
 RemoteFileMetadata = namedtuple('RemoteFileMetadata',
                                 ['filename', 'url', 'checksum'])
 
+DATASET_NAMES = {}
+
+class _DatasetRegister(type):
+    def __new__(meta, name, bases, class_dict):
+        cls = type.__new__(meta, name, bases, class_dict)
+        if not cls.NAME is None:
+            DATASET_NAMES[cls.NAME] = cls
+        return cls
+
+
+class Dataset(Bunch, metaclass=_DatasetRegister):
+    NAME = None
+    DESCRIPTION = None
+    ARCHIVE = None
+
+    def __init__(self, data_home=None, download_if_missing=True):
+        data_location = join(get_data_home(data_home=data_home),
+                             self.NAME)
+
+        if not exists(data_location):
+            makedirs(data_location)
+
+        contents = {}
+        for file_type, meta in self.ARCHIVE.items():
+            local_path = join(data_location, meta.filename)
+            contents[file_type] = local_path
+
+            if not exists(local_path):
+                if not download_if_missing:
+                    raise IOError("Data {0}={1} not found and `download_if_missing` is "
+                                  "False".format(file_type, local_path))
+                logger.info("Downloading {0}: {1} -> {2}...".format(
+                    file_type, meta.url, local_path))
+                archive_path = _fetch_remote(meta, dirname=data_location)
+
+        module_path = dirname(__file__)
+        with codecs.open(join(module_path, 'descr', self.DESCRIPTION),
+                         encoding="utf-8") as dfile:
+            contents['DESCR'] = dfile.read()
+
+
+        # finally, init the Bunch object
+        super().__init__(**contents)
+
+
+def fetch(dataset, data_home=None, download_if_missing=True):
```

Comment:
This allows `MDADATA.fetch('adk_equilibrium')`, rather than a separate function for each dataset

Reply:
There's a reason why explicit functions: tab completion and introspection. (sklearn does it and it works really well – much better than having to know the name of the dataset)

I'd like to keep explicit functions – both for ease of use and for same "look and feel" as sklearn.datasets (as well as getting docs!)

We can have a generic mechanism and generate the `fetch_*` functions.

Reply:
Yes, that's nice for this case. But did you look at some of the other accessors like `fetch_adk_transitions_DIMS` where we get a tar file and unpack? We might be able to reduce our requirements to these two types of cases.

Reply:
`RemoteFileMetadata` is verbatim from sklearn. Might be useful to keep it that way and really keep it simple. 

If anything, we should build data structures that contain `RemoteFileMetadata` instances that map remote and local. Have a look at the transitions dataset to see what else we have. 

Finally, have a look at [sklearn.datasets](https://github.com/scikit-learn/scikit-learn/tree/master/sklearn/datasets) (and the outstanding [docs](http://scikit-learn.org/stable/datasets/index.html)) to see the variance. I think one reason for copy&paste code is that ultimately each dataset in the wild might have slightly different requirements. Still, that's not to say that we can't try to get a bit of order in ;-).